### PR TITLE
Use libsolv implementation of strict repo priority

### PR DESCRIPTION
**Pull request by adriendelsalle**
Description
---

Improve handling of `strict-channel-priority` using [`libsolv` strict repo priority](https://github.com/openSUSE/libsolv/pull/459).

### CODE REVIEW

**Comment by adriendelsalle** [comment]
testing it with `create -n test xtensor -c default -c conda-forge --strict-channel-priority --no-rc -vvv` gives:
```
choice rule creation took 0 ms
15888 pkg rules, 2 * 5 update rules, 1 job rules, 0 infarch rules, 0 dup rules, 0 choice rules, 0 best rules, 0 yumobs rules
0 black rules, 0 recommends rules, 78108 repo priority rules
overall rule memory used: 2203 K
(...)
solver took 2 ms
final solver statistics: 0 problems, 0 learned rules, 0 unsolvable
solver_solve took 77 ms
```

without `--strict-channel-priority` gives:
```
choice rule creation took 0 ms
15888 pkg rules, 2 * 5 update rules, 1 job rules, 0 infarch rules, 0 dup rules, 0 choice rules, 0 best rules, 0 yumobs rules
0 black rules, 0 recommends rules, 0 repo priority rules
overall rule memory used: 372 K
(...)
solver took 1 ms
final solver statistics: 0 problems, 0 learned rules, 0 unsolvable
solver_solve took 7 ms
```
so about 70ms in solving_solve, and 2MB in memory (for the 78k rules created)

File: mamba/utils.py [hunk 1]
```diff
@@ -123,7 +123,10 @@ def load_channels(
     )
 
     if strict_priority is None:
-        strict_priority = context.channel_priority == ChannelPriority.STRICT
+        strict_priority = context.channel_priority in [
+            ChannelPriority.STRICT,
+            ChannelPriority.FLEXIBLE,
+        ]
```
Comment (adriendelsalle) [inline mamba/utils.py:130]: maybe it's preferable to call that variable `has_priority` to not be confused?

**Review by adriendelsalle — COMMENTED** [review]
LGTM

**Comment by adriendelsalle** [comment]
Adding some cache and skipping solvables that are not part of the solving gives huge performance improvements:
```
15888 pkg rules, 2 * 5 update rules, 1 job rules, 0 infarch rules, 0 dup rules, 0 choice rules, 0 best rules, 0 yumobs rules
0 black rules, 0 recommends rules, 101 repo priority rules
overall rule memory used: 375 K
(...)
solver took 2 ms
final solver statistics: 0 problems, 0 learned rules, 0 unsolvable
solver_solve took 14 ms
```
+7ms in solving_solve, and 3kB in memory (for the 101k rules created)

**Comment by wolfv** [comment]
nice! :)

**Comment by adriendelsalle** [comment]
Yeah, even for `python` with 1333 repo priority rules, it takes only +30kB memory and +20ms

**Pull request merged**